### ISSUE TITLE
`u_cache` and `du_cache`, what needs to be updated in a `DiscreteCallback`?

### ISSUE THREAD

**Issue by goretkin** [JuliaDiffEq/OrdinaryDiffEq.jl#906]
Should they be defined in: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/a0c90a790ec22e62f39244827d14dea681322736/src/misc_utils.jl#L50

I am not sure how `full_cache` is supposed to work in the following example, since what would be the `f1` field of a rate type:
https://github.com/JuliaDiffEq/DiffEqDocs.jl/blame/master/docs/src/features/diffeq_arrays.md#L173

Also not sure if it's just enough to update `integrator.u`

**Comment by goretkin** [comment]
I didn't see `u_cache` and `du_cache` in the documentation, but I thought it's what I needed since it seems like `full_cache(::Tsit5Cache)` also iterates through all the `k1`...`k7` fields of `rateType`.

```julia
@cache struct Tsit5Cache{uType,rateType,uNoUnitsType,TabType} <: OrdinaryDiffEqMutableCache
  u::uType
  uprev::uType
  k1::rateType
  k2::rateType
  k3::rateType
  k4::rateType
  k5::rateType
  k6::rateType
  k7::rateType
  utilde::uType
  tmp::uType
  atmp::uNoUnitsType
  tab::TabType
end
```

```julia
julia> import OrdinaryDiffEq

julia> c = OrdinaryDiffEq.Tsit5Cache(1.0,1.0, 2.0,2.0,2.0,2.0,2.0,2.0,2.0, 1.0, 1.0, 1.0, 99.9)
OrdinaryDiffEq.Tsit5Cache{Float64,Float64,Float64,Float64}(1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 99.9)

julia> OrdinaryDiffEq.full_cache(c)
(1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0)
```

The `rateType` fields won't have an `f1` --- that wouldn't make sense.

**Comment by goretkin** [comment]
```julia
  if isinplace(prob) && typeof(u) <: AbstractArray && eltype(u) <: Number && uBottomEltypeNoUnits == uBottomEltype # Could this be more efficient for other arrays?
    if !(typeof(u) <: ArrayPartition)
      rate_prototype = recursivecopy(u)
    else
      rate_prototype = similar(u, typeof.(oneunit.(recursive_bottom_eltype.(u.x))./oneunit(tType))...)
    end
  else
    if uBottomEltypeNoUnits == uBottomEltype
      rate_prototype = u
    else # has units!
      rate_prototype = u/oneunit(tType)
    end
  end
  rateType = typeof(rate_prototype) ## Can be different if united
```

The path that is "unit-aware" doesn't use similar. It does ` rate_prototype = u/oneunit(tType)`.

**Comment by goretkin** [comment]
Which, from what I'm gathering, is unintentional. But I think it's actually correct. That extra stuff in the `DEDataArray` should only come into play when evaluating the dynamics on a state, either to save [what I believe are called] algebraic variables or to rely on some "parameter", like a discrete-time controller.

**Comment by goretkin** [comment]
> That carries forward the array type though?

No, it doesn't appear to be the case.


```julia

mutable struct SimulationState{T, X, U} <: DEDataArray{T, 1}
    x::X        # ODE State Space
    control::U
end


julia> initial_conditions = SimulationState([0.0 * u"m", 0.0 * u"m/s"], 0.0 * u"N")
2-element SimulationState{Quantity{Float64,D,U} where U where D,Array{Quantity{Float64,D,U} where U where D,1},Quantity{Float64,𝐋*𝐌*𝐓^-2,Unitful.FreeUnits{(N,),𝐋*𝐌*𝐓^-2,nothing}}}:
      0.0 m
 0.0 m s^-1

julia> initial_conditions / 1
2-element Array{Quantity{Float64,D,U} where U where D,1}:
      0.0 m
 0.0 m s^-1

julia> initial_conditions ./ 1
2-element Array{Quantity{Float64,D,U} where U where D,1}:
      0.0 m
 0.0 m s^-1
```

**Comment by devmotion** [comment]
What would be wrong with
```julia
rate_prototype = @. u0 * $(inv(oneunit(eltype(tspan))))
```
I tried that in DelayDiffEq, and it seemed to work - but maybe I'm missing something?

**Comment by devmotion** [comment]
Would that help in the DEDataArray case?

**Comment by goretkin** [comment]
Yes, there should be a better interface. The callback should be able to change "u" in one place, since it changes at a single instant in time. I would like an interface that basically looks like this, but doesn't require me to have my own for-loop

```julia
function simulate_hybrid(continuous, discrete, u_0, t_span, Δt, parameters)
    # controller happens in discrete time. physics happens in continuous time
    (t_0, t_f) = t_span

    t = t_0
    state = u_0

    ts = [t_0]
    us = [u_0]

    while t < t_f
        t′ = min(t_f, t + Δt)

        action = discrete(state, parameters, t)

        ode_prob = ODEProblem(
            (state, parameters, time) -> continuous(state, action, parameters, time),
            state,
            (t, t′),
            parameters
        )

        sol = solve(ode_prob, Tsit5())
        @assert sol.t[end] == t′
        t = t′
        state = sol.u[end]
        push!(ts, t′)
        push!(us, state)
    end

    return (t=ts, u=us)
end
```

**Comment by goretkin** [comment]
I am using it exactly how it's being used in the docs? I thought this was a/the motivating example for `DEDataArray`.

**Comment by goretkin** [comment]
This will require a separate mechanism for saving the data, right? I rather enjoyed that my `sol.u` vector was going to include all of the information from my simulator.

A parameter that changes within the course of the simulation, in a way that depends on the simulation itself (e.g. it's not just a tape-recording, it's a controller that depends on the state) doesn't fit my usage of the word "parameter". That's not really an argument against using `parameter` for that purpose, but I don't like it.

From a more generic point of view, I'd say there is some simulation state (that's what I was calling it). Some aspects of the state change continuously, and it's the responsibility of the ODE integrator to handle that. Another aspect of the state changes discretely, and it's the `DiscreteCallback`'s responsibility to update that. So the current design of `DEDataArray` works for me in that model quite well for hybrid continuous/discrete simulations.

**Comment by ChrisRackauckas** [comment]
Then I think the best thing is to just loop through the full_cache for now and do a type check. Other iterators could come back by modifying the macro to use the pieces that are split out (you can see in the macro this is half implemented)

**Comment by goretkin** [comment]
It seems like part of the problem is that some (most? all?) methods take advantage of "FSAL", first-step-as-last. The du/dt computed at the final stage of the previous step can be reused as the first stage of the current step. But it's not possible to do this when there's a discrete callback event between two integration steps, which is why the cached need to be updated.

So there are places where a change in u invalidates the cache, but it should really only involve one evaluation at a given u, not looping through several cache variables and recomputing it.

In other words, there should be just one `u`, maybe `integrator.u_prev` or `integrator.u` for the callback to change. if a "rate" is invalidated because of that, then it should be recomputed. I haven't been able to follow where this happens now.

**Comment by ChrisRackauckas** [comment]
Indeed, that's usually what's done, but that's not the problem in your example because of the implicit state held in the array type.

**Comment by goretkin** [comment]
Sorry, I didn't understand that. Maybe you could tell me what problem I was pointing out to help me get on the same page.

I think that ideally, in the callback, I could just update `integrator.u` (which contains the implicit (maybe better: "discrete-time") state in the DEDataArray) . Instead there are several "u"-like things I need to update. Conceptually, all the "u"s should all be at the same value--there is a unique value of `u(t)` for the `t` at which the callback is triggered.

There's no way to do this currently because just changing the `integrator.u` does not update the fsalfirst value (k1, in the RK4 method). What are the other reasons it does not work?

**Comment by ChrisRackauckas** [comment]
Changing `integrator.u` is fine because it triggers the cache invalidation and recomputes the FSAL part. The problem is that `integrator.cache.tmp` needs to have a new state value, since that's what is actually shown to you during the integration.

**Comment by goretkin** [comment]
"actually shown to you during the integration" means that it's the value that is saved to the solution?

**Comment by ChrisRackauckas** [comment]
no, usually the calls in that (and some other methods) are `f(ksomething,tmp,p,t)`

**Comment by goretkin** [comment]
At least in the `RK4` case that's not true? https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/d0a8c04415f5817d536f3cd9737d4333d17adb46/src/perform_step/fixed_timestep_perform_step.jl#L320
The value of `tmp` before the `perform_step!` does not appear to matter.

**Comment by ChrisRackauckas** [comment]
The isn't `u`, it's the fact that if you only update `u` then the values in the derivatives don't change.

**Comment by goretkin** [comment]
> Changing `integrator.u` is fine because it triggers the cache invalidation and recomputes the FSAL part.

> The isn't `u`, it's the fact that if you only update `u` then the values in the derivatives don't change.

I don't know how to reconcile these two comments.

--- 
Perhaps you'll let me be more concrete. This is the structure:

```
while do_integration
  # calls `apply_step!`
  # `integrator.uprev = integrator.u`
  # `integrator.fsalfirst = integrator.fsallast`
  # update `integrator.fsalfirst` if necessary
  loopheader!(integrator)

  # u = u_prev + Δu
  perform_step!(integrator, integrator.cache)

  # callbacks happen here
  loopfooter!(integrator)
end
```

What is Δu in the case of RK4? It is `(dt/6)*(2*(k₂ + k₃) + (k₁ + k₄))`, which is available after the following block executes
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/d0a8c04415f5817d536f3cd9737d4333d17adb46/src/perform_step/fixed_timestep_perform_step.jl#L309-L319

So all it cares about there is `fsalfirst`, if I'm understanding correctly.

Just after that block:
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/d0a8c04415f5817d536f3cd9737d4333d17adb46/src/perform_step/fixed_timestep_perform_step.jl#L320

This updates `k` based on the new `u`. This is an example of "values in the derivative" that would need to be recomputed. (Though in this case, I am not sure why it's necessary since this computation is only used on the next `perform_step!`, if I'm using the term I just learned, RK4 does not have the FSAL property.)

`rk4cache.fsailfirst` is aliased to `integrator.fsalfirst`, and `rk4cache.k` is aliased to `integrator.fsallast`.

So it seems my callback, at least for RK4, would be correct if it updated
`integrator.u` and then did `f(integrator.fsalfirst, integrator.u, p, t)`, which happens in `reset_fsal!`.

There's a bit of logic for deciding if `reset_fsal!` including the flag `integrator.u_modified`. It looks like `integrator.u_modified` gets set on discrete callbacks.

So then it should be absolutely, 100%, fine to just touch `integrator.u` in my discrete callback. Yeah?

**Comment by ChrisRackauckas** [comment]
Yes, it's absolutely fine to touch `integrator.u` in a callback. In fact, it's standard to do so.

**Comment by goretkin** [comment]
I apologize, but this has been an extremely confusing thread.

> I'm not sure that should be using an array type though. How about using a parameter, and saving the array for that parameter's time series in a callback?

> Then I think the best thing is to just loop through the full_cache for now and do a type check. Other iterators could come back by modifying the macro to use the pieces that are split out (you can see in the macro this is half implemented)

Why did these suggestion come up if my use-case allows for just updating `integrator.u`?

**Comment by ChrisRackauckas** [comment]
It's required in the documentation example 

http://docs.juliadiffeq.org/latest/features/diffeq_arrays.html#Example:-A-Control-Problem-1

which I thought you were referring to with

>I am not sure how full_cache is supposed to work in the following example, since what would be the f1 field of a rate type:

since I assumed you meant `f1` to be the `f1` in that array. In that example, if you don't update `f1` in the cache variables, then it's not updated in the cache values used internally for the user's `u` and thus the parameter never updates.

**Comment by ChrisRackauckas** [comment]
We could make it a thing to only call `f` on `integrator.u` to fix this issue. Some algorithms already do this.

**Comment by goretkin** [comment]
Then I think my objection is to

> Yes, it's absolutely fine to touch integrator.u in a callback. In fact, it's standard to do so.

It's not fine, since it's not sufficient.

**Comment by ChrisRackauckas** [comment]
If it's an array which only holds integrated values, it's okay. That's the caveat. If you put extra information that isn't effected by broadcast, it's invisible to the integrator.

**Comment by goretkin** [comment]
> I'm not sure that should be using an array type though. How about using a parameter, and saving the array for that parameter's time series in a callback?

I'm coming around to this idea being the sanest approach. I understand the motivation for `DEDataArray`, but I think it ends up creating too many problems when its used for computation. I think something like `DEDataArray` is a good idea as a mechanism for storing auxiliary information in the solution.

**Comment by ChrisRackauckas** [comment]
Since these functions don't exist anymore, I'm closing this.
